apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/FlushableConsumerBase.cs

```diff
@@ -0,0 +1,113 @@
+﻿// Copyright 2017 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Diagnostics.Common
+{
+    /// <summary>
+    /// Base class for the <see cref="IFlushableConsumer{T}"/> that uses a <see cref="SemaphoreSlim"/> to
+    /// protect the flush and receive calls.
+    /// </summary>
+    internal abstract class FlushableConsumerBase<T> : IFlushableConsumer<T>
```

Comment:
At a later date, we may wish to rename this to `SynchronizedFlushableConsumerBase` or something similar, but that can definitely wait (as it's not a public name).